staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go

```diff
@@ -128,8 +127,8 @@ func (c *codec) Decode(data []byte, defaultGVK *schema.GroupVersionKind, into ru
 	// create a new instance of the type so we always exercise the conversion path (skips short-circuiting on `into == obj`)
 	decodeInto := into
 	if into != nil {
-		if _, ok := into.(runtime.Unstructured); ok && !into.GetObjectKind().GroupVersionKind().GroupVersion().Empty() {
-			decodeInto = reflect.New(reflect.TypeOf(into).Elem()).Interface().(runtime.Object)
+		if u, ok := into.(runtime.Unstructured); ok && !into.GetObjectKind().GroupVersionKind().GroupVersion().Empty() {
+			decodeInto = u.NewEmptyInstance()
```

Comment:
having paged this back into memory... I don't actually think this is correct. The goal here is to decode the content into its natural version, then convert it to the version expressed in `into` (done below in `c.convertor.Convert(obj, into, c.decodeVersion)`)

The decoder is not the thing that does conversion

Reply:
I see. Yes, it doesn't seem quite correct. It's maybe only correct for the case that `data` and `defaultGVK` both don't say anything about the type.

What about adding this at the beginning:

```
	if defaultGVK == nil {
		intoGVK := into.GetObjectKind().GroupVersionKind()
		defaultGVK = &intoGVK
	}
```

This is the code calling Decode, btw:
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/fake/client.go#L201

Please note:
- The tracker.List returns a List object without populating the GVK.
- Therefore, the json it marshals doesn't have GVK either.
- `fakeClient.List()` passes nil as default GVK into `UniversalDecoder().Decode()`

If adding that doesn't seem right, then probably controller-runtime should pass `gvk` as the defaultGVK.

Reply:
> It's maybe only correct for the case that data and defaultGVK both don't say anything about the type.

Being called like that seems like a bug to me. I wouldn't expect to compensate for that scenario inside this method.